File: validate_stack_sequences/ValidateStackSequences.java
class ValidateStackSequences {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        if (pushed == null || popped == null || pushed.length != popped.length) {
            return false;
        }
        if (pushed.length == 0) {
            return true;
        }

        Deque<Integer> stack = new ArrayDeque<>();
        int popPtr = 0;
        int pushPtr = 0;
        for (; pushPtr < pushed.length && popPtr < popped.length; pushPtr++) {
            while (!stack.isEmpty() && stack.peekFirst() == popped[popPtr]) {
                stack.pollFirst();
                popPtr++;
            }
            stack.offerFirst(pushed[pushPtr]);
        }

        while (!stack.isEmpty() && stack.peekFirst() == popped[popPtr]) {
            stack.pollFirst();
            popPtr++;
        }

        return stack.isEmpty() && popPtr == popped.length;
    }
}

File: validate_stack_sequences/Validate_Stack_Sequences.cpp
class Validate_Stack_Sequences {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        if (pushed.empty() && popped.empty()) {
            return true;
        }
        if (pushed.size() != popped.size()) {
            return false;
        }

        int push_ptr = 0;
        int popped_ptr = 0;
        stack<int> my_stack;
        for (; push_ptr < pushed.size() && popped_ptr < popped.size(); push_ptr++) {
            while (!my_stack.empty() && my_stack.top() == popped[popped_ptr]) {
                my_stack.pop();
                popped_ptr++;
            }
            my_stack.push(pushed[push_ptr]);
        }

        while (!my_stack.empty() && my_stack.top() == popped[popped_ptr]) {
            my_stack.pop();
            popped_ptr++;
        }

        return my_stack.empty() && popped_ptr == popped.size();
    }
};